apps/condo/domains/billing/access/BillingReceipt.js

```diff
@@ -5,31 +5,82 @@
 const { throwAuthenticationError } = require('@open-condo/keystone/apolloErrorFormatter')
 const { find } = require('@open-condo/keystone/schema')
 
+const { CONTEXT_FINISHED_STATUS: ACQUIRING_CONTEXT_FINISHED_STATUS } = require('@condo/domains/acquiring/constants/context')
 const { canManageBillingEntityWithContext } = require('@condo/domains/billing/utils/accessSchema')
+const { CONTEXT_FINISHED_STATUS: BILLING_CONTEXT_FINISHED_STATUS } = require('@condo/domains/miniapp/constants')
 const { RESIDENT } = require('@condo/domains/user/constants/common')
 
+// We allow only receipts:
+//    1. With matching Property (compare by addressKey)
+//    2. For organizations with Finished AcquiringContext
+//    3. With Finished BillingContext
+async function buildResidentAccessToReceiptsQuery (userId) {
+    if (!userId) {
+        return false
+    }
+    const residents = await find('Resident', { user: { id: userId }, deletedAt: null })
+    if (!residents || residents.length === 0) {
+        return false
+    }
+    let serviceConsumers = await find('ServiceConsumer', {
+        resident: { id_in: residents.map(({ id }) => id) },
+        deletedAt: null,
+    })
+    if (!serviceConsumers || serviceConsumers.length === 0) {
+        return false
+    }
+    const organizationIds = serviceConsumers.map(({ organization }) => organization )
+    // Exclude all serviceConsumers for organizations without AcquiringContext in Finished status
+    const organizationsWithContract = (await find('AcquiringIntegrationContext', {
+        organization: { id_in: organizationIds, deletedAt: null },
+        integration: { deletedAt: null },
+        status: ACQUIRING_CONTEXT_FINISHED_STATUS,
+        deletedAt: null,
+    })).map(({ organization }) => organization)
+    serviceConsumers = serviceConsumers.filter(({ organization }) => organizationsWithContract.includes(organization))
```

Comment:
List to set is O(N), set.has is O(1) * N,  so it O(N) total.
List includes is O(N) * N, so it O(N^2) total.

🤔

Reply:
You are welcome suggest a better solution